### PR TITLE
fix(gateway): isolate routes to SNI matches

### PR DESCRIPTION
**Pull request by michaelbeaumont**
Most illustrative are the changed golden files, added e2e test as well.
The MeshHTTP/TCPRoute vs MeshGatewayRoute implementations were diverging too much so this contains a lot of refactoring.

There's more refactoring to be done but I don't want to make the PR even harder to review.

Try reviewing by commit

- [x] fix MeshGatewayRoute
- [x] fix policies

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go [hunk 1]
```diff
@@ -92,23 +92,23 @@ func generateGatewayRoutes(
 	}
 
 	resources := core_xds.NewResourceSet()
-	routeConfig := plugin_gateway.GenerateRouteConfig(info)
-
 	// Make a pass over the generators for each virtual host.
-	for _, hostInfo := range hostInfos {
-		vh, err := plugin_gateway.GenerateVirtualHost(ctx, info, hostInfo.Host, hostInfo.Entries())
+	for hostname, hostInfos := range hostInfosByHostname {
+		routeConfig := plugin_gateway.GenerateRouteConfig(info.Proxy, info.Listener.Protocol, info.Listener.ResourceName+":"+hostname)
```
Comment (michaelbeaumont) [inline pkg/plugins/policies/meshhttproute/plugin/v1alpha1/gateway.go:97]: this is the core change, we create multiple RouteConfigurations

ignore the string concatenations

File: pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml [hunk 2]
```diff
@@ -729,9 +729,9 @@ Routes:
               clusters:
               - name: external-http2-httpbin-29c391043cf34cd7
                 weight: 1
-    edge-gateway:HTTP:8081:
+    edge-gateway:HTTP:8081:*:
       ignorePortInHostMatching: true
-      name: edge-gateway:HTTP:8081
+      name: edge-gateway:HTTP:8081:*
```
Comment (lahabana) [inline pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml:734]: There's no size limit on names?

**Review by jakubdyszkiewicz — APPROVED** [review]

File: pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml (hunk 2 above)
Comment (bartsmykla) [reply pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml:734]: I was not able to find any

File: test/e2e_env/kubernetes/gateway/gateway.go [hunk 3]
```diff
@@ -229,6 +310,17 @@ spec:
...
 				).To(Succeed())
+
+				Eventually(func(g Gomega) {
+					var err error
+					clusterIP, err = k8s.RunKubectlAndGetOutputE(
+						kubernetes.Cluster.GetTesting(),
+						kubernetes.Cluster.GetKubectlOptions(namespace),
+						"get", "service", "simple-gateway", "-ojsonpath={.spec.clusterIP}",
+					)
+					g.Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply test/e2e_env/kubernetes/gateway/gateway.go:321]: https://github.com/kumahq/kuma/issues/9070

File: pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml:734]: Should we add a maxlength to listener.hostname?

Comment (bartsmykla) [reply pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml:734]: Let's do it

Comment (michaelbeaumont) [reply pkg/plugins/runtime/gateway/testdata/http/cross-mesh-gateway.yaml:734]: @lahabana @jakubdyszkiewicz done, added to UPGRADE.md

**Review by bartsmykla — APPROVED** [review]

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**